### PR TITLE
Resize rbd block

### PR DESCRIPTION
**Pull request by humblec**
Fixes: #1566 
Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

### CODE REVIEW

File: internal/rbd/controllerserver.go [hunk 1]
```diff
@@ -300,6 +306,16 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 		return nil, err
 	}
 
+	// we have to resize the RBD volume with block mode request
+	if expansionRequired {
+		util.DebugLog(ctx, "\n requested volume size: %v is greater than the parent volume, increasing the size", newVolSize)
```
Comment (Madhu-1) [inline internal/rbd/controllerserver.go:311]: Why we need debug log here?

**Comment by Madhu-1** [comment]
@humblec are we resizing the rbd block device? if not, can you please update the PR title and provide few details in the PR description.

File: internal/rbd/controllerserver.go [hunk 2]
```diff
@@ -300,6 +306,16 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
...
+		err := rbdVol.resize(newVolSize)
```
Comment (Madhu-1) [inline internal/rbd/controllerserver.go:312]: what if the volume is created successfully and resize fails are we validating the new size and making sure the new requested volume has a new size?

File: internal/rbd/controllerserver.go [hunk 3]
```diff
@@ -188,31 +188,37 @@ func getGRPCErrorForCreateVolume(err error) error {
 	return status.Error(codes.Internal, err.Error())
 }
 
-// validateRequestedVolumeSize validates the request volume size with the
-// source snapshot or volume size, if there is a size mismatches it returns an error.
-func validateRequestedVolumeSize(rbdVol, parentVol *rbdVolume, rbdSnap *rbdSnapshot, cr *util.Credentials) error {
+// source snapshot or volume size, if there is a size missmatch and if the new volume is not block mode
+// it returns an error.
+func validateRequestedVolumeSize(rbdVol *rbdVolume, parentVol *rbdVolume, rbdSnap *rbdSnapshot, cr *util.Credentials, isBlock bool) (bool, error) {
```
Comment (Madhu-1) [inline internal/rbd/controllerserver.go:193]: why we need a bool flag in return here? if we return bool from here we are doing two validation, can we skip this one as do validation in the caller itself?

**Comment by nixpanic** [comment]
> /retest all

CephFS tests might be failing, rebase once #1599 is merged

File: internal/rbd/controllerserver.go [hunk 4]
```diff
@@ -90,7 +90,7 @@ func (cs *ControllerServer) validateVolumeReq(ctx context.Context, req *csi.Crea
 	return nil
 }
 
-func (cs *ControllerServer) parseVolCreateRequest(ctx context.Context, req *csi.CreateVolumeRequest) (*rbdVolume, error) {
+func (cs *ControllerServer) parseVolCreateRequest(ctx context.Context, req *csi.CreateVolumeRequest) (*rbdVolume, bool, error) {
```
Comment (nixpanic) [inline internal/rbd/controllerserver.go:93]: Please document the `bool` for return value. Users of the function do not know what the bool means.

Comment (nixpanic) [reply internal/rbd/controllerserver.go:93]: The callers of this function need to be adjusted in the same commit. This is currently not compiling, which means `git bisect` will break.

File: internal/rbd/controllerserver.go (hunk 3 above)
Comment (nixpanic) [reply internal/rbd/controllerserver.go:193]: I do not get it either. The return value needs to be explained in the comment of the function. Please follow go-doc standards too.

File: internal/rbd/controllerserver.go [hunk 5]
```diff
@@ -235,7 +235,7 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 	}
 	defer cr.DeleteCredentials()
 
-	rbdVol, err := cs.parseVolCreateRequest(ctx, req)
+	rbdVol, isBlock, err := cs.parseVolCreateRequest(ctx, req)
```
Comment (nixpanic) [inline internal/rbd/controllerserver.go:238]: Maybe introduce a `rbdVol.isBlock()` function or member instead? That prevents the need for the additional boolean return values and parameters

File: e2e/utils.go [hunk 6]
```diff
@@ -12,6 +12,8 @@ import (
 	"sync"
 	"time"
 
+	"k8s.io/apimachinery/pkg/api/resource"
+
```
Comment (nixpanic) [inline e2e/utils.go:16]: follow standard import guidelines please

**Review by nixpanic — CHANGES_REQUESTED** [review]

**Comment by humblec** [comment]
Revisit this PR>

**Pull request closed without merging**